src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,152 @@ public static Path of(URI uri) {
      */
     Path getFileName();
 
+    /**
+     * Returns the file name extension of this path as an
+     * {@code Optional<String>}.  The extension is defined to be the portion
+     * of the {@code String} representation of the file name after the last
+     * dot ('.').  If the first character in the file name string is a dot
+     * it is ignored.  If the extension cannot be determined, then an
+     * {@link Optional#empty empty} {@code Optional} is returned. This will
+     * occur if the path has zero elements ({@link #getFileName()} returns
+     * {@code null}), or the file name string does not contain a dot, or only
+     * the first character is a dot. If the last character is a dot, the
+     * extension is the empty {@code String} {@code ""}.
+     *
+     * @implSpec
+     * The default implementation is equivalent for this path to:
+     * <pre>{@code
+     *     String name = getFileName().toString();
+     *     int lastDot = name.lastIndexOf('.');
+     *     if (lastDot > 0)
+     *         lastDot == name.length() - 1) ?
+     *             Optional.of("") :
+     *             Optional.of(name.substring(lastDot + 1));
+     *     else
+     *         Optional.empty();
+     * }</pre>
+     *
+     * @return  an {@code Optional} which either contains the file name of
+     *          this path, or is {@link Optional#empty empty} if the extension
+    *           is indeterminate
```

Comment:
Can it say "not found";  indeterminate sounds non-deterministic but the algorithm is.